### DISCUSSION TITLE
How do you communicate with a bluetooth module?

### DISCUSSION THREAD

**Discussion by hankertrix** [Rahix/avr-hal#565] [Q&A]
The Arduino library seems to have a SoftwareSerial module, which I can't find in this HAL. Are there any examples of working with a Bluetooth module, like an HM-10?

I just need to be able to send text commands from a phone to the Arduino using the Bluetooth module.

**Comment by Rahix** [comment]
Hi,

you can either use a hardware serial which is going to be the most reliable solution.  That's just any of the `USART` peripherals that your chip has available.

If you really want software serial, your best bet is the [`bitbang-hal`](https://github.com/sajattack/bitbang-hal) crate.

**Reply by hankertrix** [reply]
Thanks for your reply. If I understand you correctly, this means that I should connect my Bluetooth module to a `USART` pin on my Arduino Mega 2560 board instead of a regular pin to have serial communication, and write a function to handle receiving and transmitting data.

Is the above correct? Sorry I am quite new to Arduino development.

**Reply by stappersg** [reply]
> Is the above correct?

Yes.

Mostly.

The minor correct is changing _"connect my Bluetooth module to an `USART` pin on Mega 2560"_ into _"connect my Bluetooth module to `USART` pin**s** on Mega 2560"_ ( note the plural **s** )

> I am quite new to Arduino development.

And probably also new to electronics.  And that is fine.  Every Electronics Engineer has been there.  (We all start as beginners.)   Thing I'm trying to tell is that you are learning many new things. So be ready for many initial failures. Just enjoy the learning experience.

**Reply by hankertrix** [reply]
Ah I see. Yeap I am new to electronics as well.

Thank you!
